dali/operators/image/clahe/clahe_op.cu

```diff
@@ -0,0 +1,1221 @@
+    float val = static_cast<float>(cdf[i]) * lutScale;
+    lut[i] = static_cast<uint8_t>(lrintf(fminf(fmaxf(val, 0.f), 255.f)));
+  }
+}
+
+extern "C" void LaunchMegaFusedHistClipCdfLut(const uint8_t *y_plane, int H, int W, int tiles_x,
+                                              int tiles_y, float clip_limit_rel, uint8_t *luts,
+                                              cudaStream_t stream) {
+  int tile_w = div_up(W, tiles_x);
+  int tile_h = div_up(H, tiles_y);
+  dim3 grid(tiles_x, tiles_y, 1);
+  int threads = 256;  // Optimized for occupancy
+
+  // Calculate shared memory needed
+  int warps_per_block = threads / 32;
+  size_t shmem = (warps_per_block + 2) * 256 * sizeof(unsigned int);  // warp_hists + hist + cdf
+
+  mega_fused_hist_clip_cdf_lut_kernel<<<grid, threads, shmem, stream>>>(
+      y_plane, H, W, tiles_x, tiles_y, tile_w, tile_h, clip_limit_rel, luts);
+}
+extern "C" void LaunchCLAHE_Grayscale_U8_NHWC(const uint8_t *src_gray, uint8_t *dst_gray, int H,
+                                              int W, int tiles_x, int tiles_y, float clip_limit_rel,
+                                              unsigned int *tmp_histograms,  // tiles*bins
+                                              uint8_t *tmp_luts,             // tiles*bins
+                                              cudaStream_t stream) {
+  // Use mega-fused version for larger images where the fusion overhead pays off
+  int total_tiles = tiles_x * tiles_y;
+  if (total_tiles >= 16) {  // Threshold where fusion is beneficial
+    LaunchMegaFusedHistClipCdfLut(src_gray, H, W, tiles_x, tiles_y, clip_limit_rel, tmp_luts,
+                                  stream);
+  } else {
+    // Use traditional 3-kernel approach for smaller tile counts
+    LaunchHistPerTile256(src_gray, H, W, tiles_x, tiles_y, tmp_histograms, stream);
+    LaunchClipCdfToLut256(tmp_histograms, H, W, tiles_x, tiles_y, clip_limit_rel, tmp_luts, stream);
+  }
+  LaunchApplyLUTBilinearToGray(src_gray, dst_gray, H, W, tiles_x, tiles_y, tmp_luts, stream);
+}
+
+extern "C" void LaunchCLAHE_RGB_U8_NHWC(const uint8_t *src_rgb, uint8_t *dst_rgb,
+                                        uint8_t *y_plane,  // [H*W]
+                                        int H, int W, int tiles_x, int tiles_y,
+                                        float clip_limit_rel,
+                                        unsigned int *tmp_histograms,  // tiles*bins
+                                        uint8_t *tmp_luts,             // tiles*bins
+                                        cudaStream_t stream) {
+  LaunchRGBToYUint8NHWC(src_rgb, y_plane, H, W, stream);
+  LaunchHistPerTile256(y_plane, H, W, tiles_x, tiles_y, tmp_histograms, stream);
+  LaunchClipCdfToLut256(tmp_histograms, H, W, tiles_x, tiles_y, clip_limit_rel, tmp_luts, stream);
+  LaunchApplyLUTBilinearToRGB(src_rgb, y_plane, dst_rgb, H, W, tiles_x, tiles_y, tmp_luts, stream);
+  CUDA_CHECK(cudaGetLastError());
```

Comment:
Please use `CUDA_CALL` which translates the error into a proper exception with the necessary details (see `include/dali/core/cuda_error.h`).